### PR TITLE
Limit the number of concurrent tests in integration.go

### PR DESCRIPTION
**Pull request by yujuhong**
This is to address the integration test timeout problem (#6651).

### CODE REVIEW

**Comment by yujuhong** [comment]
Waiting for travis/shippable results to see if this would drastically lengthen the overall test latency.

**Comment by vmarmol** [comment]
@yujuhong it didn't seem to really change the runtime of Travis.

File: cmd/integration/integration.go [hunk 1]
```diff
@@ -994,16 +995,26 @@ func main() {
...
-	for i := range testFuncs {
-		f := testFuncs[i]
-		go func() {
-			f(kubeClient)
-			wg.Done()
-		}()
+	// Only run at most maxConcurrency tests in parallel.
+	numFinishedTests := 0
+	for numFinishedTests < len(testFuncs) {
+		testToRun := len(testFuncs) - numFinishedTests
```
Comment (vmarmol) [inline cmd/integration/integration.go]: nit: numTestsToRun?

File: cmd/integration/integration.go [hunk 2]
```diff
@@ -994,16 +995,26 @@ func main() {
...
+		glog.Infof("Running %d tests in parallel.", testToRun)
```
Comment (vmarmol) [inline cmd/integration/integration.go]: We should move this after the if, else we may print a wrong value

**Comment by yujuhong** [comment]
Not always. In one run it had `I0409 22:34:01.362097   32622 integration.go:1002] Running 6 tests in parallel.`, instead of running all 8 tests together.

`runtime.NumCPU()` may not be a good indicator if the machine is overloaded. I just restarted shippable again to see if the number varies in different runs. If shippable times out when having >8 cpus, we'll need to set a hard limit (which will affect local runs) or pass a flag to restrict this for shippable or travis only.

File: cmd/integration/integration.go (hunk 2 above)
Comment (yujuhong) [inline cmd/integration/integration.go]: Yep, I just saw that. Thanks!

**Comment by vmarmol** [comment]
It's probably not a good indicator if the machine is overloaded. It'll just tell us the number of cores on the machine. The number of runnable cores may be less than this (if we get a cpu mask applied to us for example).

**Comment by vmarmol** [comment]
It may make sense to set this to a static value for now, I don't know how feasible it'll be for us to detect how loaded the CI machine is and change our behavior based on that.

**Comment by yujuhong** [comment]
In general, we use `runtime.NumCPU()` in many places to set `GOMAXPROCS`, so if these overloaded machines have more CPUs, we may use more threads :\

Yes, it's not feasible to detect the load on the machines, and it varies. I was suggesting that maybe we could pass a fixed number for travis/shippable configuration, but allows developers to use as many cores as their machine have locally. I'll test with a fixed number locally and measure the performance difference. If it's okay, I'll just set it to a static number.

**Comment by vmarmol** [comment]
WDYT of exposing this as a flag in the integration test? We can default to SOME_NUMBER or allow developers to raise this (with set == -1 for NumCpus()) if they want to run it faster locally.

**Comment by yujuhong** [comment]
@vmarmol, that's why I suggested, but maybe I didn't express it well :)

**Comment by vmarmol** [comment]
Ah I see it now, I'm on-board :D

**Comment by yujuhong** [comment]
Updated and passed all the tests. Only time will tell if this helps...

**Comment by vmarmol** [comment]
lol, sounds ominous :)

LGTM

File: cmd/integration/integration.go [hunk 3]
```diff
@@ -994,16 +998,26 @@ func main() {
...
-		}()
+	// Only run at most maxConcurrency tests in parallel.
+	numFinishedTests := 0
+	for numFinishedTests < len(testFuncs) {
+		numTestsToRun := len(testFuncs) - numFinishedTests
+		if maxConcurrency > 0 && numTestsToRun > maxConcurrency {
+			numTestsToRun = maxConcurrency
+		}
+		glog.Infof("Running %d tests in parallel.", numTestsToRun)
+		var wg sync.WaitGroup
```
Comment (lavalamp) [inline cmd/integration/integration.go]: The canonical way to do this in go is to start N worker threads all reading from a channel, and then dump everything you want done into the channel.

**Comment by lavalamp** [comment]
Did this actually reduce the running time? I'd expect doing things more serially to make it slower, not faster.

Also I think running everything at once is valuable for detecting races and making sure asynchronous code keeps working.

**Comment by yujuhong** [comment]
@lavalamp, thanks for the pointer. @yifan-gu has submitted a PR to improve this (using buffered channel as semaphore) yesterday in #/6669.

This goal of the PR is not to reduce the running time (latency), and it would not. It is meant to reduce the load on system since all the tests are sharing the same  apiserver, scheduler, etc (assuming shippable/travis wouldn't schedule more tasks onto one machine). If we keep adding tests, then we'll have to scale the timeout values along with it, which is a pain. I think we should have a separate performance suite similar to our e2e to track performance regression. Integration tests should focus on testing the correctness because (1) we don't control the testing infrastructure (2) it is too expensive as it runs for every PR and a lot of time and efforts are wasted in triaging the failure (3) even if we can catch *some* bugs, it is hard to reproduce the case locally (4) it makes people ignore the tests results.

That said, it is possible that there is a performance regression (or even deadlock) causing the timeout problem. The PR was to stop wasting all developers' time. I will file an issue for the potential performance problem. Feel free to chime in or revert this PR if you think having all the tests running is more valuable :)

**Pull request merged**